### PR TITLE
[WIP][RLlib] IMPALA TF Vtrace v2

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: avnish <avnish@anyscale.com>

laundry list: 

- [x] new Vtrace with tests
- [ ] Impala tf trainer with vtrace loss. Working with discrete and multi-discrete action space
- [ ] test for convergence using synchronous training loop. Will still use a replay buffer. Will need a sampling utility?
- [ ] Impala tf policy ( this needs to access the relevant fields so all the data is available at training time)
- [ ] learning rate and entropy scheduler
- [ ] integrate into training_iteration loop
- [ ] release test at scale

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by kouroshHakha — COMMENTED** [review]
Looks good, a couple of questions and nits. 
Also can you rename the impala_rl_trainer.py to a vtrace module. This is meant to be just the vtrace utility, right?

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    tensor: Union[tf.Tensor, List[tf.Tensor]],
+    trajectory_len: int = None,
+    recurrent_seq_len: int = None,
+    drop_last: bool = False,
+):
+    """Swaps batch and trajectory axis.
+
+    Args:
+        tensor: A tensor or list of tensors to swap the axis of.
+            NOTE: The tensor(s) must flattened along the batch axis.
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:17]: What if they don't?

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    if recurrent_seq_len:
+        B = tf.shape(recurrent_seq_len)[0]
+        T = tf.shape(tensor)[0] // B
+    else:
+        T = trajectory_len
+        B = tf.shape(tensor)[0] // T
+    rs = tf.reshape(tensor, tf.concat([[B, T], tf.shape(tensor)[1:]], axis=0))
+
+    # swap B and T axes
+    res = tf.transpose(rs, [1, 0] + list(range(2, 1 + int(tf.shape(tensor).shape[0]))))
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:50]: I stole this line from the existing rllib impl. lemme try this out when i get the chance, or leave a todo

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    time dimension ranging from 0 to T-1. B refers to the batch size and
+    ACTION_SPACE refers to the list of numbers each representing a number of
+    actions.
+
+    Args:
+        target_action_log_probs: Action log probs from the target policy. A float32
+            tensor of shape [T, B].
+        behaviour_action_log_probs: Action log probs from the target behavior policy. A
+            float32 tensor of shape [T, B].
+        discounts: A float32 tensor of shape [T, B] with the discount encountered
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:98]: yes but only because the discount factor should be 0 at timesteps where the episode is done/terminated

File: rllib/algorithms/impala/tf/impala_rl_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:17]: trajectories in this case will have the same length

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            make_time_major(_tensor, trajectory_len, recurrent_seq_len, drop_last)
+            for _tensor in tensor
+        ]
+
+    assert bool(trajectory_len) != bool(
+        recurrent_seq_len
+    ), "Either trajectory_len or recurrent_seq_len must be set."
+
+    if recurrent_seq_len:
+        B = tf.shape(recurrent_seq_len)[0]
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:42]: this is legacy code. I don't have an answer here or an easy way to test this without knowing how rnn sequences will be implemented with the new models.

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    # swap B and T axes
+    res = tf.transpose(rs, [1, 0] + list(range(2, 1 + int(tf.shape(tensor).shape[0]))))
+
+    if drop_last:
+        return res[:-1]
+    return res
+
+
+def get_log_rhos(target_action_log_probs, behaviour_action_log_probs):
+    """With the selected log_probs for multi-discrete actions of behaviour
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:58]: oh gotcha thx for catching this.

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    and target policies we compute the log_rhos for calculating the vtrace."""
+    t = tf.stack(target_action_log_probs)
+    b = tf.stack(behaviour_action_log_probs)
+    log_rhos = tf.reduce_sum(t - b, axis=-1)
+    return log_rhos
+
+
+def vtrace_tf2(
+    target_action_log_probs,
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py:67]: done.

File: rllib/algorithms/impala/tests/test_vtrace_v2.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
+import unittest
+import numpy as np
+
+from gymnasium.spaces import Box, Discrete
+
+from ray.rllib.algorithms.impala.tf.impala_rl_trainer import (
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tests/test_vtrace_v2.py:6]: kk

File: rllib/algorithms/impala/tests/test_vtrace.py [hunk 8]
```diff
@@ -35,14 +35,14 @@
...
 def _ground_truth_calculation(
-    vtrace,
     discounts,
     log_rhos,
     rewards,
     values,
     bootstrap_value,
     clip_rho_threshold,
     clip_pg_rho_threshold,
+    vtrace=None,
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tests/test_vtrace.py:45]: I got rid of the vtrace arg. It used to be a named tuple with the vtrace returns and likelihood adjusted advantages.

File: rllib/algorithms/impala/tests/test_vtrace.py [hunk 9]
```diff
@@ -35,14 +35,14 @@
...
 ):
     """Calculates the ground truth for V-trace in Python/Numpy."""
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tests/test_vtrace.py]: done

File: rllib/algorithms/impala/tf/impala_rl_trainer.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    recurrent_seq_len: int = None,
+    drop_last: bool = False,
+):
+    """Swaps batch and trajectory axis.
+
+    Args:
+        tensor: A tensor or list of tensors to swap the axis of.
+            NOTE: The tensor(s) must flattened along the batch axis.
+        trajectory_len: The length of the trajectory being transformed.
+        recurrent_seq_len: Sequence lengths if recurrent or None
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py]: addressed

File: rllib/algorithms/impala/tf/impala_rl_trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/impala/tf/impala_rl_trainer.py]: I got rid of this function all together, it wasn't doing enough.

**Review by kouroshHakha — APPROVED** [review]
LGTM
